File: problems/200/solution.cpp
// Accepted: 23ms (88.33%), 15.73MB (74.83%)

class Solution {
private:
    char WATER = '0';
    int DY[4] = { 0, 0, -1, 1 };
    int DX[4] = { -1, 1, 0, 0 };

    void visitIsland(vector<vector<char>>& grid, int y, int x) {
        if (!isInGrid(grid, y, x) || grid[y][x] == WATER) {
            return;
        }

        grid[y][x] = WATER;

        for (int i = 0; i < 4; ++i) {
            visitIsland(grid, y+DY[i], x+DX[i]);
        }
    }

    bool isInGrid(vector<vector<char>>& grid, int y, int x) {
        return 0 <= x && x < grid[0].size() && 0 <= y && y < grid.size();
    }

public:
    int numIslands(vector<vector<char>>& grid) {
        int count = 0;

        for (int y = 0; y < grid.size(); ++y) {
            for (int x = 0; x < grid[0].size(); ++x) {
                if (grid[y][x] == WATER) {
                    continue;
                }

                ++count;
                visitIsland(grid, y, x);
            }
        }

        return count;
    }
};

File: problems/200/solution.py
# Accepted: 279ms (20.34%), 18.96MB (53.09%)

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        WATER = "0"
        DY = [0, 0, -1, 1]
        DX = [-1, 1, 0, 0]

        def visitIsland(y: int, x: int) -> None:
            if not isInGrid(y, x) or grid[y][x] == WATER:
                return

            grid[y][x] = WATER

            for dy, dx in zip(DY, DX):
                visitIsland(y+dy, x+dx)

        def isInGrid(y, x) -> bool:
            return 0 <= x < len(grid[0]) and 0 <= y < len(grid)

        # count islands

        count = 0
        for y in range(len(grid)):
            for x in range(len(grid[0])):
                if grid[y][x] == WATER:
                    continue

                visitIsland(y, x)
                count += 1

        return count
